src/Spectre.libWavelet/DaubechiesFiltersDenoiser.cpp

```diff
@@ -0,0 +1,72 @@
+/*
+ * DaubechiesFiltersDenoiser.cpp
+ * Denoises the signal using undecimated wavelet transform.
+ *
+   Copyright 2018 Michal Gallus
+   
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+   
+   http://www.apache.org/licenses/LICENSE-2.0
+   
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+#include "DaubechiesFiltersDenoiser.h"
+#include "SoftThresholder.h"
+
+namespace spectre::algorithm::wavelet
+{
+DaubechiesFiltersDenoiser::DaubechiesFiltersDenoiser()
+{
+}
+
+static inline void ExtractCoefficientsForNoiseEstimation(Signal& highFrequencyCoeffs,
+    WaveletCoefficients& coefficients)
+{
+    CoefficientList& noiseEstimationCoefficients = coefficients.data[0][0];
+    auto chunkStart = noiseEstimationCoefficients.begin();
+    auto chunkEnd = chunkStart + highFrequencyCoeffs.size();
+    highFrequencyCoeffs.assign(chunkStart, chunkEnd);
+}
+
+static inline WaveletCoefficients TresholdSignal(const NoiseEstimator& noiseEstimator,
+    WaveletCoefficients& coefficients, size_t signalLength)
+{
+    Signal highFreqCoefficients(signalLength);
+    ExtractCoefficientsForNoiseEstimation(highFreqCoefficients, coefficients);
+    DataType noiseTreshold = noiseEstimator.Estimate(highFreqCoefficients);
+    SoftThresholder tresholder(noiseTreshold);
+    return tresholder(std::move(coefficients));
+}
+
+static inline WaveletCoefficients Decompose(const WaveletDecomposerRef& decomposer,
```

Comment:
I don't see a point for creating this and following function, but there's nothing bad about it also, so fine :D

Reply:
Yeah, it looks redundant, but the point was to keep all the code in `Denoise` on the same level of abstraction. This gets inlined anyway, so no performance is damaged.

Reply:
Yeah, that's what I saw.